crates/cranelift/src/gc/enabled.rs

```diff
@@ -56,7 +64,309 @@ pub fn gc_ref_table_fill_builtin(
     Ok(func_env.builtin_functions.table_fill_gc_ref(func))
 }
 
+pub fn translate_struct_new(
+    func_env: &mut FuncEnvironment<'_>,
+    builder: &mut FunctionBuilder<'_>,
+    struct_type_index: TypeIndex,
+    fields: &[ir::Value],
+) -> WasmResult<ir::Value> {
+    gc_compiler(func_env).alloc_struct(func_env, builder, struct_type_index, &fields)
+}
+
+fn default_value(
+    cursor: &mut FuncCursor,
+    func_env: &FuncEnvironment<'_>,
+    ty: &WasmStorageType,
+) -> ir::Value {
+    match ty {
+        WasmStorageType::I8 | WasmStorageType::I16 => cursor.ins().iconst(ir::types::I32, 0),
+        WasmStorageType::Val(v) => match v {
+            WasmValType::I32 => cursor.ins().iconst(ir::types::I32, 0),
+            WasmValType::I64 => cursor.ins().iconst(ir::types::I64, 0),
+            WasmValType::F32 => cursor.ins().f32const(0.0),
+            WasmValType::F64 => cursor.ins().f64const(0.0),
+            WasmValType::V128 => cursor.ins().iconst(ir::types::I128, 0),
+            WasmValType::Ref(r) => {
+                assert!(r.nullable);
+                let (ty, needs_stack_map) = func_env.reference_type(r.heap_type);
+
+                // NB: The collector doesn't need to know about null references.
+                let _ = needs_stack_map;
+
+                cursor.ins().iconst(ty, 0)
+            }
+        },
+    }
+}
+
+pub fn translate_struct_new_default(
+    func_env: &mut FuncEnvironment<'_>,
+    builder: &mut FunctionBuilder<'_>,
+    struct_type_index: TypeIndex,
+) -> WasmResult<ir::Value> {
+    let interned_ty = func_env.module.types[struct_type_index];
+    let struct_ty = match &func_env.types[interned_ty].composite_type {
+        WasmCompositeType::Struct(s) => s,
+        _ => unreachable!(),
+    };
+    let fields = struct_ty
+        .fields
+        .iter()
+        .map(|f| default_value(&mut builder.cursor(), func_env, &f.element_type))
+        .collect::<StructFieldsVec>();
+    gc_compiler(func_env).alloc_struct(func_env, builder, struct_type_index, &fields)
+}
+
+pub fn translate_struct_get(
+    func_env: &mut FuncEnvironment<'_>,
+    builder: &mut FunctionBuilder<'_>,
+    struct_type_index: TypeIndex,
+    field_index: u32,
+    struct_ref: ir::Value,
+) -> WasmResult<ir::Value> {
+    // TODO: If we know we have a `(ref $my_struct)` here, instead of maybe a
+    // `(ref null $my_struct)`, we could omit the `trapz`. But plumbing that
+    // type info from `wasmparser` and through to here is a bit funky.
+    builder.ins().trapz(struct_ref, ir::TrapCode::NullReference);
```

Comment:
One idea for a future trick: we could unmap the lowest page of memory in the GC heap so address 0 in the gc heap translates to a fault. That way we could push this trap into an annotated trap code on the `MemFlags` on the load below (or store for `struct.set`) which means we could, by construction, completely eliminate this trap as well. 

(possibly all on top of a cranelift optimization pass to remove the checks entirely)

Reply:
We discussed this a bit in today's cranelift meeting. In summary, this is neat but kind of conflicts with our plans of eventually using linear memories to implement the GC heap, since the pooling allocator would have to allocate a linear memory for the GC heap, then unmap the first page, give it to Wasm until the instance is done, then remap the first page before returning it to the memory pool. This is all doable of course, but the extra special-case code paths are annoying and adding more virtual memory calls in the pooling allocator is always fraught perf-wise.